### PR TITLE
v4: Restrict Flagged CNR installs on non-local listeners

### PR DESCRIPTION
**Pull request by ltdrdata**
## Purpose and scope

A successful Registry install response previously allowed a Flagged CNR version to proceed. Require Flagged versions to satisfy the same installation conditions as Active versions, plus a server listening exclusively on loopback or an explicit administrator override. Covers glob and legacy installation, switching, updates, deferred startup and server-scheduled restores, plus legacy batch reinstallation.

The allow_flagged_nodepack_install option belongs in config.ini’s [default] section. It defaults to false; only case-insensitive true enables it, after restarting ComfyUI. The override is intended for a trusted private network and permits non-loopback listeners. It does not detect private networks or bypass common security checks, Registry refusals or Git/pip restrictions.

## Key changes

- Use status from the existing CNR install response. Check policy before changing the current pack and reuse that response for execution, without another Registry lookup.
- Allow CNR release installation on non-local listeners at normal-or-lower security levels in both servers. Active/Pending releases use this common gate; Flagged releases must also pass the listener/override check. Strong security still denies all CNR installation requests, and Git/nightly retain their existing restrictions.
- Validate legacy reinstallation before removal, then install and rerun the pack’s script. Both cores preserve files and activation state on policy denial, retain the reason in queue/update results and avoid retrying rejected restore targets.
- Recheck deferred work against current startup settings and pass the resolved server permission to scheduled restore children. Standalone CLI retains its existing interface and local administrator behavior.
- Show a short security-configuration denial in the UI. Only terminal logs provide the loopback listener and private-network override remedies, including restart instructions.

Affected modules: CNR metadata mapping, installation cores, server queues, startup processing, configuration and README. Deferred switches use stored status without refreshing it from the Registry. Older reservations without status require loopback-only listeners or the override; otherwise startup rejects them without changing the pack and tells the user to request installation again. Enabling an already installed identical version remains a local activation with no Registry lookup or installation script.

## Validation approach

Core tests run production installation paths with controlled external responses. E2E uses actual HTTP requests, queues, ZIP installation, restart and CLI subprocesses, with a local Registry and isolated node paths, settings and databases.

| Intent | Observed result |
|---|---|
| Add restrictions only to Flagged | Active/Pending controls install; Flagged needs loopback or override. Strong-security controls reject both Active and Flagged before metadata lookup. |
| Preserve state on denial | Existing files, versions and activation state survive; no download or reservation is created. |
| Avoid duplicate lookups | Count metadata requests through installation and restore; deferred switches make no additional lookup. |
| Preserve execution context | Observe immediate file replacement versus deferred reservation, callback timing, standalone CLI behavior and actual server-scheduled restore children. Existing-version activation needs no Registry lookup or installation script. |
| Keep errors and settings correct | Verify short UI errors versus terminal remedies, Registry refusal, listener edge cases, old reservations and configuration persistence. |

Tests fail when status propagation is removed, Registry refusals are ignored, or reinstall script execution is suppressed. Immediate-switch checks fail when the execution flag is dropped, and missing-status reservations are tested under both allowed and denied startup settings. Restored code passes; successful Git reinstall cases cover both catalogue entries and the CNR repository fallback.

## Validation results

Current source is 0c8f423f on manager-v4, which includes #3297. Its complete Git tree is identical to the verified UI + Flagged integration 6df4aa14. Integration results below were recorded on 1a04122a; the only difference is the strengthened legacy denial-response test, and production and policy E2E files are identical. Focused policy/configuration regressions were rerun on 0c8f423f: 176 passed.

| Scope | Result |
|---|---|
| Pre-rebase standalone 0ceeec62, non-E2E regression | 453 passed; 12 skipped; 3 inherited failures |
| UI + Flagged integration 1a04122a, non-E2E regression | 569 passed; 12 skipped; the same 3 failures |
| Policy server E2E on 1a04122a, glob and legacy | 108 passed; 12 skipped |
| Live legacy browser E2E on 1a04122a | 7 scenarios passed; no uncaught browser exceptions |

The three Git backup fixture failures also occur on unchanged manager-v4. The complete local regression suite is not green. The twelve non-E2E skips require a separate CLI environment. Eight policy E2E skips require a preinstalled pack under strong security; four exclude loopback from non-local Git checks. Strong-security denial is tested directly.

Ruff, OpenAPI, code-quality and Socket checks passed on 0c8f423f. [CLI E2E CI](https://github.com/Comfy-Org/ComfyUI-Manager/actions/runs/35404887995) also passed on Ubuntu, Windows and macOS after rebasing onto the merged UI change.

Policy E2E uses a local Registry with real HTTP routes, queues, archive installation, deferred startup and CLI subprocesses. Separate Playwright scenarios exercise the integrated legacy UI, real installation and WebSocket completion, Active installation, Flagged denial and configuration preservation. Browser checks use frontend 1.45.15 and single-tab sessions; the legacy menu opens through its registered command because of an inherited toolbar issue.

Tests use Python 3.12 and ComfyUI 039ed38e on Python 3.10.20. The public Registry was not exercised in this run. Snapshot checks establish admission, extraction and denial-state preservation; fresh-snapshot install.py completion and file-lock/deferred-deletion handling remain outside this PR.

The existing deferred-deletion/state-consistency defect is tracked in [#3301](https://github.com/Comfy-Org/ComfyUI-Manager/issues/3301), including removal-result semantics, restart cleanup and regression acceptance criteria. The uninstall-failure test in this PR checks a reported failure; it does not establish correct reporting of a deferred deletion.

### CODE REVIEW

**Comment by ltdrdata** [comment]
The PR is now rebased on manager-v4 after #3300; the current head is 8bb27c00. Automated code review was skipped because the base is not the default branch, so that status is not an approval. Please review the current change, including preservation of Active installation, the additional Flagged restriction, and state preservation on denial. @coderabbitai review

File: comfyui_manager/glob/manager_core.py [hunk 1]
```diff
@@ -1549,12 +1571,12 @@ async def install_by_id(self, node_id: str, version_spec=None, channel=None, mod
         if self.is_disabled(node_id, "cnr"):
             # enable and switch version if cnr is disabled (not specified version)
             self.unified_enable(node_id, "cnr")
-            return self.cnr_switch_version(node_id, version_spec, no_deps=no_deps, return_postinstall=return_postinstall)
+            return self._cnr_switch_version(node_id, node_info, no_deps=no_deps, return_postinstall=return_postinstall)
```
[the remark this answers is not in the document]
Comment (ltdrdata) [reply comfyui_manager/glob/manager_core.py]: Confirmed and fixed in 0ceeec62 in both cores. install_by_id now forwards instant_execution through both reused-version branches. Regression checks execute the real switch paths and inspect file replacement, script execution, reservation timing and Registry call count. Four immediate-switch cases failed before the fix; all 12 immediate/deferred/fresh-install controls now pass. The corresponding v3 calls already forwarded the flag.

File: comfyui_manager/legacy/manager_core.py [hunk 2]
```diff
@@ -1538,17 +1518,89 @@ async def install_by_id(self, node_id: str, version_spec=None, channel=None, mod
...
         if self.is_enabled(node_id, "cnr"):
-            return self.cnr_switch_version(node_id, version_spec, no_deps=no_deps, return_postinstall=return_postinstall)
+            return self._cnr_switch_version(node_id, node_info, no_deps=no_deps, return_postinstall=return_postinstall)
 
-        res = self.cnr_install(node_id, version_spec, instant_execution=instant_execution, no_deps=no_deps, return_postinstall=return_postinstall)
+        res = self._cnr_install(node_id, node_info, instant_execution, no_deps, return_postinstall)
         if res.result:
             self.active_nodes[node_id] = version_spec, res.to_path
 
         return res
```
[the remark this answers is not in the document]
Comment (ltdrdata) [reply comfyui_manager/legacy/manager_core.py]: Confirmed and fixed in 0ceeec62, together with the glob finding. Both reused-version calls now forward instant_execution. The tests observe immediate file replacement and script execution versus deferred reservation, while requiring exactly one Registry request. The four affected cases failed before correction and pass afterward, with deferred and fresh-install controls retained.

File: comfyui_manager/legacy/manager_core.py [hunk 3]
```diff
@@ -1468,64 +1482,30 @@ async def install_by_id(self, node_id: str, version_spec=None, channel=None, mod
...
-                elif version_spec == 'nightly':
-                    cnr_utils.generate_cnr_id(to_path, node_id)
-                    self.active_nodes[node_id] = 'nightly', to_path
-            else:
-                return res
-
-            return res.with_target(version_spec)
+        # Check before changing the existing pack, then reuse this response.
+        try:
+            node_info = self._get_cnr_install_info(node_id, version_spec)
```
[the remark this answers is not in the document]
Comment (ltdrdata) [reply comfyui_manager/legacy/manager_core.py:1505]: The project owner has clarified that this policy governs installation, not reactivation of an already installed identical version. We are intentionally preserving this branch: it only enables the existing files and does not download or run installation scripts. Moving the Registry lookup ahead of it would add a network dependency and broaden the agreed policy.

The README now states this boundary. New regression controls exercise actual disabled-to-active movement and the subsequent active-version skip, requiring zero Registry lookups and zero installation-script calls in both cores. Reinstallation and switching to a different version remain subject to preflight policy before modifying the installed pack.

File: comfyui_manager/prestartup_script.py [hunk 4]
```diff
@@ -724,8 +740,9 @@ def execute_lazy_delete(path):
                         execute_lazy_install_script(script[0], script[2])
 
                     elif script[1] == "#LAZY-CNR-SWITCH-SCRIPT":
-                        execute_lazy_cnr_switch(script[0], script[2], script[3], script[4], script[5], script[6])
-                        execute_lazy_install_script(script[3], script[7])
+                        status = script[8] if len(script) > 8 else ''
+                        if execute_lazy_cnr_switch(script[0], script[2], script[3], script[4], script[5], script[6], status):
+                            execute_lazy_install_script(script[3], script[7])
```
[the remark this answers is not in the document]
Comment (ltdrdata) [reply comfyui_manager/prestartup_script.py:748]: The concern is valid; the owner chose a narrower compatibility rule than rejecting every old reservation. Fixed in 0ceeec62 and in v3 33c5a8c4: when stored status is missing, a non-loopback listener without the override cannot execute the reservation. The terminal tells the user to request installation again so the Registry status can be checked. Existing files remain untouched.

Loopback-only or explicitly opted-in environments can still execute these older reservations. The deferred-switch tests cover both formats, current listener/override settings, download and file effects, and no additional Registry request. README documents the migration behavior.

File: tests/e2e/test_e2e_flagged_nodepacks.py [hunk 5]
```diff
@@ -0,0 +1,494 @@
...
+                    process.wait(timeout=10)
+                except subprocess.TimeoutExpired:
+                    process.kill()
+                    process.wait(timeout=5)
+
+
+@pytest.mark.parametrize('operation', ['install', 'restore-snapshot'])
+def test_direct_cli_installs_flagged_without_server_options(tmp_path, registry, operation):
+    with running_server(tmp_path, 'legacy', '0.0.0.0', False, 'normal', registry) as server:
+        env = server.env
```
[the remark this answers is not in the document]
Comment (ltdrdata) [reply tests/e2e/test_e2e_flagged_nodepacks.py]: Accepted and fixed in 0ceeec62. Shared environment preparation now serves both the server harness and direct CLI cases. The direct CLI checks no longer start or wait for ComfyUI; both install and restore-snapshot subprocess cases pass using only the prepared isolated environment.

**Comment by ltdrdata** [comment]
Updated to 0ceeec62. Immediate switching and old-reservation handling are corrected, and the direct CLI tests no longer start ComfyUI. The project owner explicitly kept activation of an already installed identical version outside the installation policy; the README and regression controls document that boundary. Old reservations without status are denied only on non-loopback listeners without override. Every finding has a response.

The latest automatic status says “review skipped” on this non-default base. @coderabbitai review

File: comfyui_manager/legacy/manager_core.py [hunk 6]
```diff
@@ -1538,17 +1518,89 @@ async def install_by_id(self, node_id: str, version_spec=None, channel=None, mod
...
+                return ManagedResult('install').fail(str(exc))
+        else:
+            try:
+                node_info = self._get_cnr_install_info(node_id, version_spec)
+            except PermissionError as exc:
+                return ManagedResult('install-cnr').fail(str(exc))
+            if node_info is None or not node_info.download_url:
+                return ManagedResult('install-cnr').fail(f'not available node: {node_id}@{version_spec}')
+
+        removed = self.unified_uninstall(node_id, version_spec == 'unknown')
```
[the remark this answers is not in the document]
Comment (ltdrdata) [reply comfyui_manager/legacy/manager_core.py:1590]: The deferred-deletion defect is valid, but it predates this PR. I compared the current legacy unified_uninstall and try_rmtree against manager-v4 a0d89e9a: both function ASTs are identical. The previous reinstall queue also used unified_uninstall before installation. The owner explicitly deferred file-lock/deferred-deletion semantics to a separate patch, so changing the removal contract here would exceed the agreed scope. This PR preserves the policy boundary: admission is checked before removal, and an uninstall that reports failure prevents installation. It does not claim that deferred deletion is completed removal. The PR's validation limits now explicitly retain this limitation; this finding remains a follow-up rather than a fix claimed by the Flagged policy PR.

File: tests/e2e/test_e2e_flagged_nodepacks.py [hunk 7]
```diff
@@ -0,0 +1,509 @@
...
+        log = (tmp_path / 'server.log').read_text()
+        calls = registry.calls[before:]
+        assert 'Restore snapshot done.' in log, log[-6000:]
+        assert 'Restore snapshot failed.' not in log, log[-6000:]
+        assert 'Error in sitecustomize' not in log, log[-6000:]
+        assert not reservation.exists()
+        assert sum('/install' in call for call in calls) == 1, (calls, log[-6000:])
+        assert any(call.startswith('/zip/') for call in calls) is allowed, calls
+        if allowed:
+            assert f'version = "{version}"' in (installed / 'pyproject.toml').read_text(), log[-6000:]
```
[the remark this answers is not in the document]
Comment (ltdrdata) [reply tests/e2e/test_e2e_flagged_nodepacks.py:501]: The proposed assertion is appropriate for validating complete snapshot restoration, but fresh-snapshot install.py execution is a known pre-existing defect intentionally split from this PR at the owner's request. It is covered by the independent fix/v4-snapshot-postinstall branch and its marker-based Red/Green tests; neither the Flagged source nor the UI + Flagged integration includes that fix. Adding the marker assertion here would require importing the separate fix. This test establishes the Flagged policy boundary through the actual restore child: one metadata lookup, allowed archive extraction, and no download or existing-state change on denial. The PR body explicitly limits snapshot validation to those observations. Normal installation/reinstallation tests separately verify their install.py markers. I am retaining this scope and keeping full snapshot postinstall verification with the independent fix.

File: comfyui_manager/legacy/manager_core.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (ltdrdata) [reply comfyui_manager/legacy/manager_core.py:1590]: The follow-up is now concretely tracked in #3301, and linked from the PR's validation limits.

I also verified behavior, rather than relying only on the unchanged function diff. Using the actual removal, reservation-writer and installation functions from base a0d89e9a and head 0ceeec62, with temporary directories and controlled filesystem failures, both revisions give identical results in all three cases: successful deletion, successful rename with deferred trash cleanup, and deletion plus rename failure. In the last case, the original contents remain, uninstall incorrectly reports success and removes the cache entry, and installation rejects the existing path. This confirms the defect and its pre-existing scope; it is not a claim of real Windows lock E2E coverage.

#3301 requires truthful removal results, state retention for surviving paths, stopping reinstall before download/scripts, preserving the restart-deletion fallback, and ensuring trash cleanup cannot remove a replacement. It includes success controls and restart validation. The PR body now explicitly distinguishes the existing test of a reported uninstall failure from this deferred-deletion defect.

Does this address the remaining concern for #3299 without requiring the deferred-deletion fix in the Flagged policy patch? If so, please resolve this thread as a tracked follow-up. If there is a remaining behavior introduced by this PR that makes that separation insufficient, please identify the path and expected result so it can be addressed before any merge. The PR remains open and unmerged.

**Comment by ltdrdata** [comment]
@coderabbitai Please assess the technical disposition of the remaining [deferred-deletion finding](https://github.com/Comfy-Org/ComfyUI-Manager/pull/3299#discussion_r4051158926).

The follow-up is now [#3301](https://github.com/Comfy-Org/ComfyUI-Manager/issues/3301). In addition to the unchanged function comparison, production removal/reservation/installation functions from base a0d89e9a and current head 0ceeec62 were exercised with temporary paths and controlled filesystem failures. Both revisions behave identically for successful deletion, successful rename with deferred trash cleanup, and deletion plus rename failure. The last case confirms the existing inconsistency: original files remain, the cache entry is removed, uninstall reports success, and installation rejects the occupied path.

The follow-up specifies truthful removal outcomes, retained state for surviving paths, preserving deferred uninstall, preventing reinstall work while the path remains, safe trash cleanup, success controls and restart validation. The PR body now links that issue and explicitly limits its uninstall-failure test to failures actually reported by uninstall.

Is this separation sufficient for the Flagged policy PR, with no deferred-deletion change required in #3299? Please state whether you agree with the scope and evidence, or identify any remaining behavior introduced by this PR that requires a correction here. The owner requested an explicit consensus on this remaining concern.

File: comfyui_manager/legacy/manager_core.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (ltdrdata) [reply comfyui_manager/legacy/manager_core.py:1590]: Consensus is confirmed in [the reviewer assessment](https://github.com/Comfy-Org/ComfyUI-Manager/pull/3299#issuecomment-5737136693): the defect is valid and pre-existing, #3301 contains the agreed separate-fix scope and acceptance criteria, and no deferred-deletion correction is required in #3299. Closing this review discussion as a tracked follow-up; #3301 remains open and the defect is not claimed fixed. #3299 remains unmerged.

**Comment by ltdrdata** [comment]
@coderabbitai review

Please review current head 0c8f423f524e01b615dfffedc662b96e482c6b62 after rebasing on manager-v4, where #3297 has now merged. The only conflicts were in the two config writers; the resolution retains the shared writer and includes the Flagged option in its owned keys, exactly as in the previously verified UI + Flagged integration. The entire resulting Git tree equals integration 6df4aa14 (tree baa2e263116749e8e5c3efb393d086cec5e963a7). Focused policy/configuration regressions passed all 176 cases. Existing policy/server/browser E2E evidence therefore applies to identical production files. No new feature change was introduced; the agreed #3301 follow-up scope remains unchanged.

File: comfyui_manager/glob/manager_core.py [hunk 8]
```diff
@@ -1290,10 +1290,24 @@ def cnr_install(self, node_id: str, version_spec=None, instant_execution=False,
...
+        except PermissionError as exc:
+            return result.fail(str(exc))
         if node_info is None or not node_info.download_url:
             return result.fail(f'not available node: {node_id}@{version_spec}')
 
+        result = self._cnr_install(node_id, node_info, instant_execution, no_deps, return_postinstall)
+        # Keep the requested version in the public result.
+        if result.target is not None:
+            result.target = version_spec
+        return result
```
[the remark this answers is not in the document]
Comment (ltdrdata) [reply comfyui_manager/glob/manager_core.py:1304]: This is intentional preservation of the public cnr_install result, rather than a new loss of information introduced by this PR.

On the unchanged manager-v4 base 4ad94b20, both public cnr_install implementations assign result.target directly from the requested version_spec. They do not replace version_spec with the Registry response version, so an omitted argument already produces target=None. The new private _cnr_install uses the checked response's concrete version, while the public wrapper preserves the previous caller-visible result semantics. The adjacent comment records this purpose.

The existing test_direct_cnr_execution_and_postinstall explicitly covers the omitted-version public call in both cores, with immediate and returned-postinstall modes, and asserts target=None for that installation call. It also verifies successful extraction, script/callback execution, exactly one metadata request and one download. Those cases passed in the 176-case rebase verification. The same test separately preserves the existing concrete target for immediate version switching.

The proposed guard would change that existing result behavior and break the compatibility control, without addressing a Flagged policy regression. I recommend retaining the current implementation in this PR. Please reconsider the finding against the base behavior and identify a specific caller regression, if one remains; changing the public result semantics should be an intentional separate API decision.
